contracts/Verification.sol

```diff
@@ -1,100 +1,88 @@
-
-    struct Tree {
-        bytes32 root;
-        uint256 numbLeaves;
+        return commitment == _commitment;
     }
 
-    Tree tree;
-    mapping(uint256 => bytes32) public leaves;
-    mapping(bytes32 => bytes32) public leafProof;
+    /**
+     * @notice Verify all elements in a Merkle Tree data structure
+     * @dev Performs an in-place merkle tree root calculation
+     * @dev Note there is currently an assumption here that if the number
+     *      of leaves is odd, the final leaf will be duplicated prior to
+     *      calling this function
+     *
+     * @param _data the array of data elements to be verified
+     * @param _commitment the expected merkle root of the structure
+     * @return a boolean value representing the success or failure of the operation
+     */
+    function verifyMerkleAll(
+        bytes32[] memory _data,
+        bytes32 _commitment
+    ) public pure returns (bool) {
+        uint256 hashLength = _data.length;
+
+        for (uint256 j = 0; hashLength > 1; j = 0) {
+            for (uint256 i = 0; i < hashLength; i = i + 2) {
+                _data[j] = keccak256(abi.encodePacked(_data[i], _data[i + 1]));
+                j = j + 1;
+            }
+            // This effectively halves the list length every time,
+            // but a subtraction op-code costs less
+            hashLength = hashLength - j;
+        }
 
-    constructor() {
-        tree = Tree("", 0);
+        return _data[0] == _commitment;
     }
 
-    function verifyDataB(bytes32[] memory _data, bytes32 _commitment)
-        public
-        returns (bool)
-    {
-
-        for(uint i = 0; i < _data.length; i++) {
-            // Insert the leaf's data into the tree
-            tree.numbLeaves = tree.numbLeaves + 1;
-            leaves[tree.numbLeaves] = _data[i];
-
-            // Hash the leaf with the current root to get the new root
-            bytes32 rootHash = keccak256(abi.encodePacked(tree.root, _data[i]));
-            // Set the root and save it as the leaf's local merkle proof
-            leafProof[_data[i]] = rootHash;
-            tree.root = rootHash;
+    /**
+     * @notice Verify a single leaf element in a Merkle Tree
+     * @dev For sake of simplifying the verification algorithm,
+     *      we make an assumption that the proof elements are sorted
+     *
+     * @param root the root of the merkle tree
+     * @param leaf the leaf which needs to be proved
+     * @param proof the array of proofs to help verify the leafs membership
+     * @return a boolean value representing the success or failure of the operation
+     */
+    function verifyMerkleLeaf(
+        bytes32 root,
+        bytes32 leaf,
+        bytes32[] calldata proof
+    ) public pure returns (bool) {
+        bytes32 computedHash = leaf;
+
+        for (uint256 i = 0; i < proof.length; i++) {
+            bytes32 proofElement = proof[i];
+
+            if (computedHash < proofElement) {
```

Comment:
What's the gas cost of this check compared to adding a bitfield to tell this function whether an element should be left or right in each hash?

Reply:
I'd say slightly less, given that the current implementation would be:
(If `n` is _number of elements_)
`n * (less_than_opcode + equality_operator) = 6n`
If you used a bitfield to encode the left/right property for each element, you would only require a single conditional operator:
`n * equality_operator = 3n`
However the bitfield itself would require an additional function parameter of type `bytes` which would needed to be loaded onto the stack from calldata, which costs roughly:
`(n * calldataload)/32`

Reply:
Just had a thought, the last calculation is actually `(n * calldataload)/256`, which makes the bitfield cost a bit less than I originally thought